clash-cores/src/Clash/Cores/Internal/UART.hs

```diff
@@ -0,0 +1,317 @@
+   -- | Wait until we are at the center of the start bit
+   | UartRxStartBit
+       { uartRxSampleCounter :: Unsigned 4
+       -- ^ Current supersample
+       }
+   -- | Receive data bits
+   | UartRxDataBits
+       { uartRxWord :: BitVector 8
+       -- ^ The data received
+       , uartRxSampleCounter :: Unsigned 4
+       -- ^ Current supersample
+       }
+   -- | Check the stop bit is correct
+   | UartRxStopBit
+       { uartRxWord :: BitVector 8
+       -- ^ The received word
+       , uartRxSampleCounter :: Unsigned 4
+       -- ^ Current supersample
+       }
+  deriving (Show, Generic, NFDataX)
+
+-- | State transition function for the UART receiver
+uartRxT
+  :: UartRxState
+  -- ^ The current state
+  -> Bit
+  -- ^ The uart receive line
+  -> (UartRxState, Maybe (BitVector 8))
+  -- ^ The next state and a potential word that has been received
+uartRxT UartRxIdle 1 = (UartRxIdle, Nothing)
+uartRxT UartRxIdle 0 = (UartRxStartBit 1, Nothing)
+uartRxT UartRxStartBit{..} b = (nextState, Nothing)
+  where
+    middleSample = uartRxSampleCounter == 8
+    stillLow = b == 0
+    validStartBit = middleSample && stillLow
+    nextCounter = uartRxSampleCounter + 1
+
+    nextState | validStartBit  = UartRxDataBits 0b1000_0000 nextCounter
+              | middleSample   = UartRxIdle
+              | otherwise      = UartRxStartBit nextCounter
+uartRxT UartRxDataBits{..} b = (nextState, Nothing)
+  where
+    middleSample = uartRxSampleCounter == 8
+    lastBit = bitToBool $ lsb uartRxWord
+    finishedReceive = middleSample && lastBit
+    nextCounter = uartRxSampleCounter + 1
+    nextWord = replaceBit (7 :: Index 8) b (shiftR uartRxWord 1)
+
+    nextState | finishedReceive = UartRxStopBit nextWord nextCounter
+              | middleSample    = UartRxDataBits nextWord nextCounter
+              | otherwise       = UartRxDataBits uartRxWord nextCounter
+uartRxT UartRxStopBit{..} b = (nextState, out)
+  where
+    middleSample = uartRxSampleCounter == 8
+    validStopBit = middleSample && b == 1
+    nextCounter = uartRxSampleCounter + 1
+    out | validStopBit = Just uartRxWord
+        | otherwise    = Nothing
+
+    nextState | middleSample = UartRxIdle
+              | otherwise    = UartRxStopBit uartRxWord nextCounter
+uartRxT _ _ = (UartRxIdle, Nothing)
+
+-- | The UART receiver.
+-- This version requires the baud rate generator to be passed in. Create one
+-- using `Clash.Cores.UART.baudrateGenerator`.
+uartRxNoBaudrateGen
+  :: HiddenClockResetEnable dom
+  => BaudrateGenerator dom
+  -- ^ Baudrate generator
+  -> Signal dom Bit
+  -- ^ The UART receive line
+  -> Signal dom (Maybe (BitVector 8))
+  -- ^ The received words
+uartRxNoBaudrateGen (BaudrateGenerator enable) b = register Nothing out
+  where
+    out = mux (isRising False $ isJust <$> fsmOut) fsmOut (pure Nothing)
+    (uartS, fsmOut) = unbundle
+                        $ regEn (UartRxIdle, Nothing) enable
+                        $ uartRxT <$> uartS <*> supersampled
```

Comment:
```suggestion
    (uartS, fsmOut) =
      unbundle $ uartRxT <$> regEn UartRxIdle enable uartS <*> supersampled
```

`fsmOut` is getting a lot of registers, let's drop one since it doesn't seem very useful.

I tried the same in `uartTxNoBaudrateGen`, but it was being obstinate. The extra register on `ackOut`  there would have been much more difficult to eliminate, and I wonder if it is at all possible to reduce the size of the state, so it can stay.